libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp

```diff
@@ -0,0 +1,232 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/distributed_learning/distributed_learning_client.hpp"
+#include "ml/optimisation/adam_optimiser.hpp"
+#include "word2vec_training_params.hpp"
+
+std::string ReadFile(std::string const &path)
+{
+  std::ifstream t(path);
+  if (t.fail())
+  {
+    throw std::runtime_error("Cannot open file " + path);
+  }
+
+  return std::string((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+}
+
+template <class TensorType>
+class Word2VecClient : public TrainingClient<TensorType>
+{
+  using DataType         = typename TensorType::Type;
+  using SizeType         = typename TensorType::SizeType;
+  using VectorTensorType = std::vector<TensorType>;
+
+public:
+  Word2VecClient(std::string const &id, TrainingParams<TensorType> const &tp,
+                 std::string const &vocab_file, SizeType batch_size, SizeType number_of_peers);
+  void PrepareModel() override;
+  void PrepareDataLoader() override;
+  void PrepareOptimiser() override;
+  void Test(DataType &test_loss) override;
+
```

Comment:
I just want to clarify that this client is currently missing several important features that should be included in a future PR. For example these two:
- SetVocab(vocab_object) // overwrites existing vocab object
- SetVocabWord(String word, SizeType word_idx, SizeType count = numeric_max<SizeType>()) // update/re-assign a particular word to a particular index (and count if needed) - internally should remove previous assignment

Reply:
Yes, these features are missing and will be included in future PR.